### DISCUSSION TITLE
How to set up Renovate for a Python repo on a private GCP artifact registry

### DISCUSSION THREAD

**Discussion by lazarillo** [renovatebot/renovate#24326] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Include searches for package dependencies on private, Google Artifact Registry Python packages using `pyproject.toml` for dependency list and a Google service account / token.

### What you tried so far.

I actually wrote a very detailed question in Stack overflow, then realized I should have probably started here first.

Here is the stack overflow question:

https://stackoverflow.com/q/77066978/534238

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by viceice** [comment]
use `_json_key_base64` as `username` and the base64 JSON value as `password` instead of `token`

**Reply by lazarillo** [reply]
Yeah, I have tried that.  It isn't _quite clear_ what is meant by the "base64 JSON value".  As shown in the link above, I discovered that Google actually uses the following _as the password_ (base64 decoded):

```json
{
  "type": "service_account",
  "project_id": "<my-project>",
  "private_key_id": "<my-private-key-id>",
  "private_key": "-----BEGIN PRIVATE KEY-----\n <stripped for security reasons> \n-----END PRIVATE KEY-----\n",
  "client_email": "my-service-account@my-project.iam.gserviceaccount.com",
  "client_id": "<removed for privacy>",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/my-service-account%40my-project.iam.gserviceaccount.com",
  "universe_domain": "googleapis.com"
}
```

So it's unclear whether I should give this whole thing (base64 encoded) as the password, just the private key (base64 encoded) as the password, or if there is some other weird configuration to try.  I have tried the entire blurb and just the private key.  Neither worked.

Have you ever successfully had renovate with a Google Cloud Python Artifact Registry that is connecting using a service account?

Thanks so much for the help!

**Comment by lazarillo** [comment]
I have a follow up, which I can instead convert to a bug, except that I am not _quite_ sure how to report it.

After further adjusting and a few PRs today, I was able to get things to work by doing the following:

1. I used the base64-encoded JSON key + password again.
2. I used the _entire_ GCP key as the password (which includes much more than just the password, more details below).
3. **I turned off encryption of the password**

Once I got this to work again, I turned the encryption back on **and it broke again**.  So I have narrowed the problem to be with the encryption.

# Details of working solution

## `renovate.json5` file:

```json
"hostRules": [
    {
       "matchHost": "us-west1-python.pkg.dev",
      "hostType": "pypi",
      "username": "_json_key_base64",
      "password": "<filtered for security, but some details below>"
    }
]
```

The password is _exactly the same base64-encoded path that is given to GCP_.  Ie, if you do something like `gcloud artifacts print-settings python`, it will output something like:

```toml
# Insert the following snippet into your .pypirc

[distutils]
index-servers =
    my-python-packages

[my-python-packages]
repository: https://us-west1-python.pkg.dev/<my-project>/<my-repo>/
username: _json_key_base64
password: <exact same password as given in hostRules password above>
```

The "exact same password" mentioned above is actually not _just_ a password.  If you decode it (`echo <password> | base64 --decode`), you will get something like this:

```json
{
  "type": "service_account",
  "project_id": "<my-project>",
  "private_key_id": "<stripped for security>",
  "private_key": "-----BEGIN PRIVATE KEY-----\n <different private key, but also base64 encoded>  \n",
  "client_email": "<my-service-account>@<my-project>.iam.gserviceaccount.com",
  "client_id": "<also stripped>",
  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
  "token_uri": "https://oauth2.googleapis.com/token",
  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/<my-service-account>%40<my-project>.iam.gserviceaccount.com",
  "universe_domain": "googleapis.com"
}
```

It was the _entire base64-encoded message_ that was successful.  Again, this is whatever is returned from the `gcloud artifacts print-settings python` call.

# Details of failed solution which should work

The only difference between this solution and the solution above is that the `"password"` was first encrypted, as per all encryptions we've been doing across our other projects.

Specifically, we go to [this page](https://app.renovatebot.com/encrypt), then enter the following:

1. Organization: "us-west1-python.pkg.dev"
2. Repository: I've tried both leaving this empty, since it is optional, and putting in the repo path to the private GCP artifact registry repo
3. Raw value: <the entire base64-encoded message/password mentioned above, which was successful>

Then the encrypted value is generated, and copied into the JSON file below:

## `renovate.json5` file:

```json
"hostRules": [
    {
        "matchHost": "us-west1-python.pkg.dev",
        "hostType": "pypi",
        "username": "_json_key_base64",
        "encrypted": {
            "password": <the encrypted value spit out by the steps described above>
        }
    }
]
```

# Debug output from the *success* run (without encrypting the password)

Renovate has a _lot_ of output, so this filters down to just the minimal parts of the output that seem relevant.  More can be added if needed:

```text
"hostRules": [
    {
      "matchHost": "us-west1-python.pkg.dev",
      "hostType": "pypi",
      "username": "_json_key_base64",
      "password": "***********"
    }
  ]

  DEBUG: Adding password authentication for us-west1-python.pkg.dev to hostRules
  DEBUG: Using file match: (^|/)pyproject\.toml$ for manager pep621
  DEBUG: Using file match: (^|/)pyproject\.toml$ for manager poetry
  DEBUG: Matched 1 file(s) for manager pep621: data/phxfakedata/pyproject.toml
  DEBUG: Matched 1 file(s) for manager poetry: data/phxfakedata/pyproject.toml
  DEBUG: Poetry: error parsing pyproject.toml
  {
    "packageFile": "data/phxfakedata/pyproject.toml"
    "err": {
      "message": "Schema error",
      "stack": "ZodError: Schema error\n    at Object.get error [as error] (/opt/containerbase/tools/renovate/36.83.0/node_modules/zod/lib/types.js:43:31)\n    at fromZodResult (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/util/result.ts:54:67)\n    at Function.parse (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/util/result.ts:549:12)\n    at Object.extractPackageFile (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/modules/manager/poetry/extract.ts:18:36)\n    at extractPackageFile (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/modules/manager/index.ts:75:9)\n    at getManagerPackageFiles (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/extract/manager-files.ts:45:43)\n    at /opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/extract/index.ts:63:28\n    at async Promise.all (index 12)\n    at extractAllDependencies (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/extract/index.ts:60:26)\n    at extract (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/process/extract-update.ts:140:28)\n    at extractDependencies (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/process/index.ts:154:26)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/index.ts:62:9)\n    at attributes.repository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/renovate.ts:18:22",
      "issues": {
        "tool": {
          "poetry": "Required"
        }
      }
    }
  }
  DEBUG: GET https://pypi.org/pypi/phxprotobuf/ = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=691)

  "pep621": [
    {
      "extractedConstraints": {
        "python": ">=3.8"
      },
      "deps": [
        {
          "packageName": "phxprotobuf",
          "depName": "phxprotobuf",
          "datasource": "pypi",
          "depType": "project.dependencies",
          "currentValue": ">=1.6.0",
          "registryUrls": [
            "https://pypi.org/pypi/",
            "https://**redacted**@us-west1-python.pkg.dev/<our-gcp-project>/<our-python-package-repo>/simple/"
          ],
          "updates": [],
          "versioning": "pep440",
          "warnings": [],
          "registryUrl": "https://**redacted**@us-west1-python.pkg.dev/<our-gcp-project>/<our-python-package-repo>/simple/",
          "currentVersion": "1.6.0"
        },
      ],
      "packageFile": "data/phxfakedata/pyproject.toml"
    }
<OTHER PACKAGES, TOO, BUT THEY ARE REMOVED FOR SOME BREVITY>
  ],

  {
    "branchName": "renovate/lock-file-maintenance-pep621",
    "prNo": null,
    "prTitle": "deps: refresh pep621 lock files",
    "result": "not-scheduled",
    "upgrades": [
      {
        "displayPending": "",
        "packageFile": "data/phxfakedata/pyproject.toml",
        "updateType": "lockFileMaintenance"
      }
    ]
  }

  DEBUG: http statistics
  {
      "https://%24%7BPYTHON_REPO_USERNAME%7D:%24%7BPYTHON_REPO_PASSWORD%7D@us-west1-python.pkg.dev/<our-gcp-project>/<our-python-package-repo>/simple/phxprotobuf/ (GET,200)": 1,  
      "https://pypi.org/pypi/phxprotobuf/ (GET,404)": 1,
      "https://pypi.org/pypi/phxprotobuf/json (GET,404)": 1,
    }
}
```

The only error captured there is the Poetry error.  This is fine because I use PEP 621 instead of Poetry.  (I'm using [PDM](https://pdm.fming.dev/latest/).)  I don't know why the Poetry checker runs, since I don't ask for it to run, but that's fine, because there is also a 200 HTTP for grabbing the proper package

# Debug output from the *failed* run (identical to above, but with encryption added)

Again, just a small portion of the error log, because Renovate is quite verbose.

```text
    "hostRules": [
      {
        "matchHost": "us-west1-python.pkg.dev",
        "hostType": "pypi",
        "username": "_json_key_base64",
        "encrypted": {
          "password": "***********"
        }
      }
    ]


DEBUG: Trying to decrypt password
DEBUG: Decrypted config using openpgp
DEBUG: Secret is scoped to a different org
{
  "orgPrefixes": [
    "US-WEST1-PYTHON.PKG.DEV/"
  ]
}

WARN: Could not parse decrypted string
{
  "err": {
    "validationError": "Encrypted secret is scoped to a different org: \"US-WEST1-PYTHON.PKG.DEV/\".",
    "message": "config-validation",
    "stack": "Error: config-validation\n    at tryDecrypt (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/config/decrypt.ts:142:31)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at decryptConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/config/decrypt.ts:186:30)\n    at decryptConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/config/decrypt.ts:238:13)\n    at mergeRenovateConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/merge.ts:252:27)\n    at getRepoConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/config.ts:12:12)\n    at initRepo (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/index.ts:44:12)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/index.ts:55:14)\n    at attributes.repository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/renovate.ts:18:22"
  }
}

DEBUG: Trying to decrypt with old private key
DEBUG: Could not decrypt using default padding
DEBUG: Could not decrypt using PKCS1 padding
INFO: Repository has invalid config
{
  "error": {
    "validationError": "Failed to decrypt field password. Please re-encrypt and try again.",
    "message": "config-validation",
    "stack": "Error: config-validation\n    at decryptConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/config/decrypt.ts:192:27)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at decryptConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/config/decrypt.ts:238:13)\n    at mergeRenovateConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/merge.ts:252:27)\n    at getRepoConfig (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/config.ts:12:12)\n    at initRepo (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/init/index.ts:44:12)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/repository/index.ts:55:14)\n    at attributes.repository (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/36.83.0/node_modules/renovate/lib/renovate.ts:18:22"
  }
}

DEBUG: raiseConfigWarningIssue()
```

Other related output is identical.

# My _guess_ to the issue

I suspect that encryption / decryption method is somehow expecting that it is _only_ decrypting a **password**, not an entire JSON message.  I'm not sure why this should matter, but I'm not sure what else could be the problem.

Regardless, I have managed to debug to the point that it is definitely the encryption portion that is failing, since I have a success doing everything the same, right up to the point of encryption.

**Reply by viceice** [reply]
you need to pass the GitHub organization and the GitHub repo (optional) to the encryption form

**Reply by rarkins** [reply]
@lazarillo I think you've misunderstood the purpose of the organizations field.

```
DEBUG: Secret is scoped to a different org
{
  "orgPrefixes": [
    "US-WEST1-PYTHON.PKG.DEV/"
  ]
}
```

The orgs in that encrypt form are meant to be the github orgs you scope the secret to

**Reply by lazarillo** [reply]
OK, thanks!  I'm not sure what you mean, though.

Can you give an example of what the `"orgPrefix"` should look like?

For instance, we are Cellpoint Digital, we use GitHub and GCP, our GitHub org path is `"https://github.com/cellpointdigital"`, and I think folks want to keep our GCP projects and folders secret, but let's just say our GCP project is called `"cellpoint"`.  Are you saying that I would use `"orgPrefixes": ["github.com/cellpointdigital"]`?  I thought I tried that, but I can try again.

**Reply by rarkins** [reply]
No, just cellpointdigital

**Comment by xcambar** [comment]
I have a similar setup, only with JFrog Artifactory instead of GCP and after many self-directed exploration of the configuration and documentation of Renovate, I can confirm that either:
* the docs for setting up a matchHost could use some clarification/simplification, since they were quite confusing to me as is.
* the decryption of the encrypted token/password does not work.

To be proven wrong, I'd like to be able to take inspiration from an available, working, example, but the most official documentation I've read, starting with:
* https://github.com/renovatebot/renovate/blob/main/docs/usage/getting-started/private-packages.md
* https://docs.mend.io/bundle/wsk/page/how_to_set_up_hostrules_for_npm_private_registries_hosted_in_jfrog_artifactory.html

have not led to a satisfying conclusion.

This discussion by @lazarillo actually offered a working option by suggesting to not use the encrypted fields.

**Comment by norman-zon** [comment]
I experience a similar issue. 
When setting the password verbatim like 

```json
{
  "hostType": "pypi",
  "matchHost": "europe-west3-python.pkg.dev",
  "username": "_json_key_base64",
  "password": "<BASE64_PW>"
}
```

renovate succesfully authenticates, but when trying to read the password from the container ENV like so

```json
"password": "process.env.GAR_PYTHON_SA_B64"
```

authentication fails with ` statusCode=401`
